File: leetcode/AIcoding/1.Gemini-2.5/go_solutions/3245-alternating-groups-iii.go
func numberOfAlternatingGroups(colors []int, k int) int {
	n := len(colors)

	// Create an array to mark where alternating conditions are met.
	// ones[i] = 1 if colors[i] != colors[(i+1)%n], else 0.
	// This captures the alternating property between adjacent elements in the circular array.
	ones := make([]int, n)
	for i := 0; i < n; i++ {
		if colors[i] != colors[(i+1)%n] {
			ones[i] = 1
		} else {
			ones[i] = 0
		}
	}

	// Extend the ones array to handle circularity for the sliding window.
	// An alternating group of length k requires k-1 alternating conditions to be true.
	// For example, for [c0, c1, c2], we need c0!=c1 and c1!=c2 (2 conditions for k=3).
	// So, we need to find windows of size k-1 in the 'ones' array that sum to k-1.
	// We extend by k-1 elements to avoid modulo operations when checking windows that wrap around.
	extendedOnes := make([]int, n+k-1)
	for i := 0; i < n+k-1; i++ {
		extendedOnes[i] = ones[i%n]
	}

	count := 0
	currentSum := 0

	// Calculate the initial sum for the first window of size k-1.
	// This window corresponds to the group starting at index 0.
	// If k=1, k-1=0, this loop does not run, currentSum remains 0.
	for i := 0; i < k-1; i++ {
		currentSum += extendedOnes[i]
	}

	// Check if the first window forms an alternating group.
	// If k=1, k-1=0, currentSum=0, so 0 == 0 is true, count becomes 1.
	if currentSum == k-1 {
		count++
	}

	// Slide the window for the remaining n-1 starting positions.
	// The window slides from index i to i+k-2 in the extendedOnes array.
	// 'i' here represents the starting index of the group in the original 'colors' array.
	for i := 1; i < n; i++ {
		// Remove the element leaving the window (extendedOnes[i-1])
		// Add the element entering the window (extendedOnes[i + k - 2])
		currentSum = currentSum - extendedOnes[i-1] + extendedOnes[i+k-2]
		if currentSum == k-1 {
			count++
		}
	}

	return count
}

File: leetcode/AIcoding/1.Gemini-2.5/c_solutions/3245-alternating-groups-iii.c
#include <stdlib.h> // For malloc and free

int numberOfAlternatingGroups(int* colors, int colorsSize, int k) {
    // Handle the special case where k is 1.
    // Any single element forms an alternating group of length 1.
    if (k == 1) {
        return colorsSize;
    }

    // Step 1: Create the 'diff' array.
    // diff[i] = 1 if colors[i] != colors[(i+1)%colorsSize], else 0.
    // This array indicates if the pair (colors[i], colors[(i+1)%colorsSize]) is alternating.
    // A group of k elements is alternating if k-1 consecutive pairs are alternating.
    int* diff = (int*)malloc(colorsSize * sizeof(int));
    // If malloc fails, return 0 (though LeetCode typically guarantees successful allocation for typical constraints).
    if (diff == NULL) {
        return 0;
    }

    for (int i = 0; i < colorsSize; i++) {
        diff[i] = (colors[i] != colors[(i + 1) % colorsSize]) ? 1 : 0;
    }

    // Step 2: Handle the case where k-1 (the required length of consecutive 1s in diff)
    // is greater than or equal to colorsSize.
    // If k-1 >= colorsSize, it means the alternating condition must hold for all n pairs in the circular array.
    // Thus, all elements in the 'diff' array must be 1.
    // If all are 1, then any starting position forms a valid group (n groups). Otherwise, 0 groups.
    if (k - 1 >= colorsSize) {
        int all_ones = 1;
        for (int i = 0; i < colorsSize; i++) {
            if (diff[i] == 0) {
                all_ones = 0;
                break;
            }
        }
        free(diff); // Free the allocated memory for diff
        return all_ones ? colorsSize : 0;
    }

    // Step 3: Use a sliding window for the case where k-1 < colorsSize.
    // We need to find windows of length k-1 in the 'diff' array (circularly) that sum to k-1.
    // To handle circularity, we extend the 'diff' array by appending its first k-1 elements.
    int extended_diff_size = colorsSize + (k - 1);
    int* extended_diff = (int*)malloc(extended_diff_size * sizeof(int));
    if (extended_diff == NULL) {
        free(diff); // Free diff if extended_diff fails
        return 0;
    }

    // Copy original diff elements
    for (int i = 0; i < colorsSize; i++) {
        extended_diff[i] = diff[i];
    }
    // Append first k-1 elements to handle circular wrap-around for windows
    for (int i = 0; i < k - 1; i++) {
        extended_diff[colorsSize + i] = diff[i];
    }
    free(diff); // The original diff array is no longer needed

    int ans = 0;
    int current_ones = 0;

    // Initialize the sum for the first window of length k-1
    // This window corresponds to the group starting at colors[0]
    for (int i = 0; i < k - 1; i++) {
        current_ones += extended_diff[i];
    }

    // Check if the first window is all ones
    if (current_ones == k - 1) {
        ans++;
    }

    // Slide the window across the extended_diff array
    // Each iteration corresponds to a new starting position for a group in the original 'colors' array
    for (int i = 1; i < colorsSize; i++) {
        // Remove the element leaving the window
        current_ones -= extended_diff[i - 1];
        // Add the new element entering the window
        current_ones += extended_diff[i + k - 2];

        // If the current window sum is k-1, it means all elements in this window are 1
        if (current_ones == k - 1) {
            ans++;
        }
    }

    free(extended_diff); // Free the allocated memory for extended_diff
    return ans;
}